File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n=s.length();
        Stack<Character> st= new Stack<>();
        for(int i=0; i< n; i++){
            char ch=s.charAt(i);
            if(ch=='('){
                st.push(')');
            }
            else if(ch=='{'){
                st.push('}');
            }
            else if(ch=='['){
                st.push(']');
            }
            else{
                if(!st.isEmpty() && st.peek()== ch){
                    st.pop();
                }
                else{
                    return false;
                }
            }
        }

        if(!st.isEmpty()) return false;


        return true;
        
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st=[]

        for  c in s:
            if c=='(':
                st.append(')')
            elif c=='{':
                st.append('}')
            elif c=='[':
                st.append(']')
            elif not st or st[-1]!=c:
                return False
            else:
                st.pop()
            
        return not st
